br/pkg/restore/import_retry.go

```diff
@@ -0,0 +1,196 @@
+	metaClient SplitClient
+
+	errors error
+	rs     retryStatus
+}
+
+func overRegionsInRange(start, end []byte, metaClient SplitClient, retryStatus retryStatus) overRegionsInRangeController {
+	return overRegionsInRangeController{
+		start:      start,
+		end:        end,
+		metaClient: metaClient,
+		rs:         retryStatus,
+	}
+}
+
+func (o *overRegionsInRangeController) OnError(ctx context.Context, result rpcResult, region *RegionInfo) {
+	o.errors = multierr.Append(o.errors, errors.Annotatef(&result, "execute over region %v failed", region.Region))
+}
+
+func (o *overRegionsInRangeController) Run(ctx context.Context, f regionFunc) error {
+	if !o.rs.ShouldRetry() {
+		return o.errors
+	}
+	tctx, cancel := context.WithTimeout(ctx, importScanRegionTime)
+	defer cancel()
+	// Scan regions covered by the file range
+	regionInfos, errScanRegion := PaginateScanRegion(
+		tctx, o.metaClient, o.start, o.end, ScanRegionPaginationLimit)
+	if errScanRegion != nil {
+		return errors.Trace(errScanRegion)
+	}
+
+	// Try to download and ingest the file in every region
+	lctx := logutil.ContextWithField(
+		ctx,
+		logutil.Key("startKey", o.start),
+		logutil.Key("endKey", o.end),
+	)
+
+	for _, region := range regionInfos {
+		result := f(lctx, region)
+
+		if !result.OK() {
+			o.OnError(ctx, result, region)
+			switch result.StrategyForRetry() {
+			case giveUp:
+				logutil.CL(ctx).Warn("unexpected error, should stop to retry", logutil.ShortError(&result), logutil.Region(region.Region))
+				return o.errors
+			case fromThisRegion:
+				logutil.CL(ctx).Warn("retry for region", logutil.Region(region.Region), logutil.ShortError(&result))
+				time.Sleep(o.rs.ExponentialBackoff())
+				continue
+			case fromStart:
+				logutil.CL(ctx).Warn("retry for execution over regions", logutil.ShortError(&result))
+				// TODO: make a backoffer considering more about the error info,
+				//       instead of ingore the result and retry.
+				time.Sleep(o.rs.ExponentialBackoff())
+				return o.Run(ctx, f)
+			}
+		}
+	}
+	return nil
+}
+
+// rpcResult is the result after executing some RPCs to TiKV.
+type rpcResult struct {
+	Err        error
+	StoreError *errorpb.Error
+}
+
+func rpcResultFromError(err error) rpcResult {
+	return rpcResult{
+		Err: err,
+	}
+}
+
+func rpcResultOK() rpcResult {
+	return rpcResult{}
+}
+
+type retryStrategy int
+
+const (
+	giveUp retryStrategy = iota
+	fromThisRegion
+	fromStart
+)
+
+func (r *rpcResult) StrategyForRetry() retryStrategy {
+	if r.Err != nil {
+		return r.StrategyForRetryGoError()
+	}
+	return r.StrategyForRetryStoreError()
+}
+
+func (r *rpcResult) StrategyForRetryStoreError() retryStrategy {
+	if r.StoreError == nil {
+		return giveUp
+	}
+
+	if r.StoreError.GetServerIsBusy() != nil ||
+		r.StoreError.GetNotLeader() != nil ||
+		r.StoreError.GetRegionNotFound() != nil {
+		return fromThisRegion
+	}
+
+	return fromStart
+}
+
+func (r *rpcResult) StrategyForRetryGoError() retryStrategy {
+	if r.Err == nil {
```

Comment:
these seems unreachable. It can be removed.

Reply:
This would be called at `StrategyForRetry`.